src/app/server-cluster/OptionalAttributes.h

```diff
@@ -0,0 +1,139 @@
+};
+
+} // namespace Internal
+
+/// It is very common that a class has optional attributes. Such optional attributes
+/// need checking for and also affect what attributes are being returned by
+/// server cluster implementations
+///
+/// This class implements the common case where attribute IDs can fit
+/// within a 32-bit bitset as most attributes start with low IDs and are incremented
+/// by one.
+///
+/// The implementation of the class generally is a wrapper over a bitset with a
+/// `IsSet()` method. Configurations should use the SupportedAttributes<...> class.
+class AttributeSet
+{
+public:
+    AttributeSet()                                       = default;
+    AttributeSet(const AttributeSet & other)             = default;
+    AttributeSet(AttributeSet && other)                  = default;
+    AttributeSet & operator=(const AttributeSet & other) = default;
+    AttributeSet & operator=(AttributeSet && other)      = default;
+
+    constexpr bool IsSet(AttributeId id) const { return (mSetBits & (1u << id)) != 0; }
+
+    /// Exposes a "force attribute bit set" without extra validation,
+    /// so that clusters can enforce specific bits to be set.
+    ///
+    /// This is NOT intended as a generic set, use `SupportedAttributes` to configure values.
+    template <AttributeId id>
+    constexpr AttributeSet & ForceSet()
+    {
+        static_assert(id < 32, "Attribute ID must be settable");
+        return Set(id, true);
+    }
+
+protected:
+    constexpr AttributeSet & Set(AttributeId id, bool value = true)
+    {
+        if (value)
+        {
+            mSetBits |= static_cast<uint32_t>((static_cast<uint32_t>(1) << id));
+        }
+        else
+        {
+            mSetBits &= ~static_cast<uint32_t>((static_cast<uint32_t>(1) << id));
+        }
+        return *this;
+    }
+
+private:
+    uint32_t mSetBits = 0;
+};
+
+/// A specialization of AttributeSet that provides checked calls to `Set`.
+///
+/// Specifically it requires that attributes are declared as part of the template
+/// parameter pack.
+///
+/// Example usage:
+///
+///    namespace chip::app::Clusters::GeneralDiagnostics {
+///
+///    using SupportedAttributes = chip::app::SupportedAttributes<
+///        Attributes::TotalOperationalHours::Id,
+///        Attributes::BootReason::Id,
+///        Attributes::ActiveHardwareFaults::Id
+///    >;
+///
+///    } // namespace chip::app::Clusters::GeneralDiagnostics
+///
+/// After this, one can:
+///
+///   GeneralDiagnostics::SupportedAttributes()
+///      .Set<GeneralDiagnostics::Attributes::TotalOperationalHours::Id>()
+///      .Set<GeneralDiagnostics::Attributes::BootReason::Id>();
+///
+/// Cluster implementations can then store a
+///   Constructor(const GeneralDiagnostics::SupportedAttributes& enabled) : mEnabledAttributes(enabled) {...}
+///
+/// where:
+///   const AttributeSet mEnabledAttributes;
+template <AttributeId... OptionalAttributeIds>
+class SupportedAttributes : public AttributeSet
```

Comment:
So a few things.

The class is called SupportedAttributes but arguments of this type generally get called `enabledAttributes`....  It gets pretty confusing, especially because these are typically not _all_ the supported attributes when used (e.g. it misses the mandatory ones).

So it seems to me that this should be called something like "OptionalAttributes" or "EnabledOptionalAttributes" or something?

Reply:
Updated naming and standardized everyting on `OptionalAttributeSet` (updated class name, header name as well as members/variables to match)